runtime/src/main/java/org/corfudb/runtime/view/LayoutView.java

```diff
@@ -126,66 +124,64 @@ public Layout prepare(long epoch, long rank)
-                        } catch (Exception e) {
-                            return null;
-                        }
-                    })
-                    .filter(x -> x != null)
-                    .toArray(LayoutPrepareResponse[]::new);
-
-            log.debug("prepare: Successful responses={}, needed={}, timeouts={}, "
-                            + "wrongEpochRejected={}",
-                    acceptList.length, getQuorumNumber(), timeouts, wrongEpochRejected);
-
-            if (acceptList.length >= getQuorumNumber()) {
-                break;
+        // when quorum is unreachable, check if outranks is majority and throw OutrankedException
+        if (promises < getQuorumNumber()) {
+            log.debug("prepare: Quorum unreachable, promises={}, required={}, timeouts={}, " +
+                            "wrongEpochRejected={}, outranks={}",
+                    promises, getQuorumNumber(), timeouts, wrongEpochRejected, outranks);
+            if (outranks >= getQuorumNumber()) {
+                throw new OutrankedException(outrankedRecord);
+            } else {
+                throw new QuorumUnreachableException(Math.toIntExact(promises), getQuorumNumber());
             }
         }
-        // Return any layouts that have been proposed before.
-        List<LayoutPrepareResponse> list = Arrays.stream(acceptList)
-                .filter(x -> x.getLayout() != null)
+
+        log.debug("prepare: Successful responses={}, needed={}, timeouts={}, " +
+                        "wrongEpochRejected={}, outranks={}",
+                promises, getQuorumNumber(), timeouts, wrongEpochRejected, outranks);
+
+        // Collect any layouts that have been proposed before.
+        List<LayoutPrepareResponse> acceptedLayouts = stream(prepareList)
+                .filter(cf -> cf.isDone() && !cf.isCompletedExceptionally())
+                .map(cf -> cf.getNow(null))
+                .filter(x -> x != null && x.getLayout() != null)
                 .collect(Collectors.toList());
-        if (list.isEmpty()) {
+
+        // Return accepted layout with the highest rank or null.
+        if (acceptedLayouts.isEmpty()) {
             return null;
         } else {
             // Choose the layout with the highest rank proposed before.
-            long highestReturnedRank = Long.MIN_VALUE;
+            long highestRank = Long.MIN_VALUE;
             Layout layoutWithHighestRank = null;
-
-            for (LayoutPrepareResponse layoutPrepareResponse : list) {
-                if (layoutPrepareResponse.getRank() > highestReturnedRank) {
-                    highestReturnedRank = layoutPrepareResponse.getRank();
+            for (LayoutPrepareResponse layoutPrepareResponse : acceptedLayouts) {
+                if (layoutPrepareResponse.getRank() > highestRank) {
+                    highestRank = layoutPrepareResponse.getRank();
                     layoutWithHighestRank = layoutPrepareResponse.getLayout();
                 }
```

Comment:
You can replace this logic with `TreeSet<Layout> layouts = new TreeSet<>(Layout.LAYOUT_COMPARATOR);` and then `layouts.last()` will return the largest LayoutPrepareResponse

Reply:
Done.